### PR TITLE
[itsy-bitsy,ztd-cuneicode,ztd-encoding-tables,ztd-idk,ztd-platform,ztd-static-containers] add new port

### PR DESCRIPTION
**Pull request by jiayuehua**
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is in the new port's versions file.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Review by LilyWangLL — DISMISSED** [review]

**Comment by jiayuehua** [comment]
I am not sure https://github.com/jiayuehua/vcpkg/blob/ity-bitsy/ports/ztd-cuneicode/fix-c32_state-alignment.patch modification is correct, althouth the change can pass CI. So I will wait for @ThePhD response in https://github.com/soasis/idk/issues/12. After His response, I will proceed the pr.

**Comment by dg0yt** [comment]
> I am not sure https://github.com/jiayuehua/vcpkg/blob/ity-bitsy/ports/ztd-cuneicode/fix-c32_state-alignment.patch modification is correct, althouth the change can pass CI. So I will wait for @ThePhD response in [soasis/idk#12](https://github.com/soasis/idk/issues/12). After His response, I will proceed the pr.

It passes CI at build time. But the errors are more likely to appear at runtime.

**Comment by dg0yt** [comment]
Before asking other questions: Are the Cmake config changes from upstream? It makes a difference for review.

**Comment by jiayuehua** [comment]
all cmake config change are make by me.

**Comment by ThePhD** [comment]
Hi, sorry for taking a while here.

The CMake changes are beyond me and I have no idea what's going on there, or why vcpkg needs them. I pushed a latest for idk, text, and cuneicode. itsy.bitsy should be good-to-go as is. All of the changes requested in this thread (asides from the cmake changes) are in latest.

Let me know if you need anything else from me,

ThePhD

**Comment by dg0yt** [comment]
The cmake change may be trimmed a little bit, and some part may be specific to vcpkg installation layout requirements.
But a key change which should be considered upstream is to install exported CMake config. (ATM the generated config is left in the build tree.) And as part of this config installation process, there should be conscious decision on the package name and on the target namespace prefix (if not omitted). vcpkg considers upstream names as official, and everything else is marked (prefixed) as unofficial.

**Comment by jiayuehua** [comment]
@LilyWangLL I have update to upstream lastest change and add unofficial namespace, but x64_osx CI likely not work, keep reporting "##[error]Failed to create CoreCLR, HRESULT: 0x80004005
,##[error]We stopped hearing from agent vcpkg-eg-mac-08. Verify the agent machine is running and has a healthy network connection. Anything that terminates an agent process, starves it for CPU, or blocks its network access can cause this error. For more information, see: https://go.microsoft.com/fwlink/?linkid=846610".

**Review by LilyWangLL — DISMISSED** [review]

File: ports/itsy-bitsy/fix-cmake.patch [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ # informs about the project, gives a description, version and MOST IMPORTANTLY
+@@ -100,10 +97,7 @@ option(ITSY_BITSY_EXAMPLES "Enable build of examples" OFF)
+ 
+ # # Dependencies
+ # ztd.idk
+-FetchContent_Declare(ztd.idk
+-	GIT_REPOSITORY https://github.com/soasis/idk.git
+-	GIT_TAG main)
+-FetchContent_MakeAvailable(ztd.idk)
++find_package(ztd.idk CONFIG REQUIRED)
```
Comment (dg0yt) [inline ports/itsy-bitsy/fix-cmake.patch:27]:
```suggestion
+find_package(ztd.idk CONFIG REQUIRED)
+add_library(ztd::itk ALIAS unofficial-ztd::ztd.idk)
```

File: ports/itsy-bitsy/fix-cmake.patch [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ 
+ # # itsy.bitsy Target
+ # Build Target
+@@ -115,7 +109,7 @@ add_library(itsy.bitsy INTERFACE)
+ add_library(itsy::bitsy ALIAS itsy.bitsy)
+ target_link_libraries(itsy.bitsy
+ 	INTERFACE
+-	ztd::idk)
++	unofficial-ztd::ztd.idk)
```
Comment (dg0yt) [inline ports/itsy-bitsy/fix-cmake.patch:36]:
```suggestion
 	ztd::idk)
```

File: ports/itsy-bitsy/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(PACKAGE_NAME itsy.bitsy CONFIG_PATH "lib/cmake/itsy.bitsy")
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")
+
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
+
+#file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```
Comment (dg0yt) [inline ports/itsy-bitsy/portfile.cmake:27]: Unneeded.

**Review by dg0yt — COMMENTED** [review]
I commented only on the first new port now, but it is similar for all ports.

And I still wonder if we should or shouldn't use the `unofficial` prefix for packages and targets. AFAICS upstream exports config and uses official namespaces, but lacks installation of the config files.

File: ports/itsy-bitsy/fix-cmake.patch [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ # Includes a bunch of basic flags and utilities shared across projects
+ # See more at the github repository link below
+ include(FetchContent)
+-FetchContent_Declare(ztd.cmake
+-	GIT_REPOSITORY https://github.com/soasis/cmake
+-	GIT_TAG main)
+-FetchContent_MakeAvailable(ztd.cmake)
+-set(CMAKE_PROJECT_INCLUDE ${ZTD_CMAKE_PROJECT_PRELUDE})
++include(CMakePackageConfigHelpers)
++include(GNUInstallDirs)
```
[the remark this answers is not in the document]
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: include(CMakePackageConfigHelpers) is needed for configure_package_config_file call.
include(GNUInstallDirs) is needed for `CMAKE_INSTALL_INCLUDEDIR` https://github.com/ThePhD/itsy_bitsy/blob/d5b6bf9509bb2dff6235452d427f0b1c349d5f8b/CMakeLists.txt#L121

File: ports/itsy-bitsy/fix-cmake.patch [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
++++ b/cmake/itsy.bitsy-config.cmake.in
+@@ -9,9 +9,12 @@
+ #  See https://github.com/ThePhD/itsy_bitsy#using-the-library for documentation.
+ 
+ @PACKAGE_INIT@
++include(CMakeFindDependencyMacro)
++find_dependency(ztd.idk CONFIG)
++include("${CMAKE_CURRENT_LIST_DIR}/itsy.bitsy-targets.cmake")
+ 
+-include("${CMAKE_CURRENT_LIST_DIR}/sol2-targets.cmake")
```
[the remark this answers is not in the document]
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: I think its a copy-paste error, thephd has another project named sol2 https://github.com/ThePhD/sol2/blob/9c882a28fdb6f4ad79a53a4191b43ce48a661175/cmake/sol2-config.cmake.in#L25

File: ports/itsy-bitsy/fix-cmake.patch (hunk 1 above)
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: have removed the unofficial-ztd namespace

File: ports/itsy-bitsy/fix-cmake.patch (hunk 2 above)
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: have remove unofficial-ztd namespace

File: ports/itsy-bitsy/portfile.cmake (hunk 3 above)
Comment (jiayuehua) [reply ports/itsy-bitsy/portfile.cmake]: have deleted this line

File: ports/itsy-bitsy/fix-cmake.patch [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+ 	DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/itsy.bitsy")
+ 
+-# pkg-config support, except on Windows
+-if(NOT WIN32 OR NOT CMAKE_HOST_SYSTEM_NAME MATCHES Windows)
+-	set(PKGCONFIG_INSTALL_DIR
+-		"${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig"
+-		CACHE PATH "Path where itsy.bitsy.pc is installed")
+-	
+-	configure_file("${CMAKE_CURRENT_SOURCE_DIR}/cmake/itsy.bitsy.pc.in" "${CMAKE_CURRENT_BINARY_DIR}/itsy.bitsy.pc" @ONLY)
+-	install(FILES "${CMAKE_CURRENT_BINARY_DIR}/itsy.bitsy.pc" DESTINATION "${PKGCONFIG_INSTALL_DIR}")
```
[the remark this answers is not in the document]
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: have reverted the change

File: ports/itsy-bitsy/fix-cmake.patch [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ 
+ # # itsy.bitsy Target
+ # Build Target
+@@ -135,7 +129,11 @@ write_basic_package_version_file(
+ export(TARGETS itsy.bitsy FILE "${CMAKE_CURRENT_BINARY_DIR}/cmake/itsy.bitsy-targets.cmake")
+ 
+ install(TARGETS itsy.bitsy
+-	EXPORT itsy.bitsy)
++	EXPORT itsy.bitsy-targets)
++install(EXPORT itsy.bitsy-targets
```
[the remark this answers is not in the document]
Comment (jiayuehua) [reply ports/itsy-bitsy/fix-cmake.patch]: have reverted the change

**Review by LilyWangLL — APPROVED** [review]

**Pull request merged**